In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline 

# Data

In [2]:
# Read the online file by the URL provides above, and assign it to variable "df"
other_path = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/auto.csv"
df = pd.read_csv(other_path, header=None)
df

0    1            2       3      4     5            6    7      8   \
0     3    ?  alfa-romero     gas    std   two  convertible  rwd  front   
1     3    ?  alfa-romero     gas    std   two  convertible  rwd  front   
2     1    ?  alfa-romero     gas    std   two    hatchback  rwd  front   
3     2  164         audi     gas    std  four        sedan  fwd  front   
4     2  164         audi     gas    std  four        sedan  4wd  front   
..   ..  ...          ...     ...    ...   ...          ...  ...    ...   
200  -1   95        volvo     gas    std  four        sedan  rwd  front   
201  -1   95        volvo     gas  turbo  four        sedan  rwd  front   
202  -1   95        volvo     gas    std  four        sedan  rwd  front   
203  -1   95        volvo  diesel  turbo  four        sedan  rwd  front   
204  -1   95        volvo     gas  turbo  four        sedan  rwd  front   

        9   ...   16    17    18    19    20   21    22  23  24     25  
0     88.6  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1     88.6  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2     94.5  ...  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3     99.8  ...  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4     99.4  ...  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
..     ...  ...  ...   ...   ...   ...   ...  ...   ...  ..  ..    ...  
200  109.1  ...  141  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
201  109.1  ...  141  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
202  109.1  ...  173  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
203  109.1  ...  145   idi  3.01  3.40  23.0  106  4800  26  27  22470  
204  109.1  ...  141  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[205 rows x 26 columns]

In [3]:
# create headers list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


In [4]:
# replace headers with correct column names
df.columns = headers
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22      ?  

[10 rows x 26 columns]

In [5]:
# save new dataframe to csv file
df.to_csv("automobile.csv", index=False)

In [6]:
# checking data types of dataframe
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


In [7]:
# describe all the columns with basic stats on numerical data  
df.describe(include = "all")

symboling normalized-losses    make fuel-type aspiration  \
count   205.000000               205     205       205        205   
unique         NaN                52      22         2          2   
top            NaN                 ?  toyota       gas        std   
freq           NaN                41      32       185        168   
mean      0.834146               NaN     NaN       NaN        NaN   
std       1.245307               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           205        205          205             205  205.000000  ...   
unique            3          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            114         96          120             202         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.756585  ...   
std             NaN        NaN          NaN             NaN    6.021776  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    205.000000          205   205     205        205.000000        205   
unique          NaN            8    39      37               NaN         60   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           94    23      20               NaN         19   
mean     126.907317          NaN   NaN     NaN         10.142537        NaN   
std       41.642693          NaN   NaN     NaN          3.972040        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        205  205.000000  205.000000   205  
unique        24         NaN         NaN   187  
top         5500         NaN         NaN     ?  
freq          37         NaN         NaN     4  
mean         NaN   25.219512   30.751220   NaN  
std          NaN    6.542142    6.886443   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

In [8]:
df

symboling normalized-losses         make fuel-type aspiration  \
0            3                 ?  alfa-romero       gas        std   
1            3                 ?  alfa-romero       gas        std   
2            1                 ?  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

## Deal with missing data

In [9]:
# replace "?" to NaN
df.replace("?", np.nan, inplace = True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [10]:
# Function to count missing values in each column (True values are missing values)
def find_missing_values(dataframe):
    missing_data = dataframe.isnull()

    for column in missing_data.columns.values.tolist():
        true_count = missing_data[column][missing_data[column] == True].count()
        if true_count != 0:
            print(f"In column {column} there are {true_count} missing data points")
            print("\n")
        
find_missing_values(df)

In column normalized-losses there are 41 missing data points


In column num-of-doors there are 2 missing data points


In column bore there are 4 missing data points


In column stroke there are 4 missing data points


In column horsepower there are 2 missing data points


In column peak-rpm there are 2 missing data points


In column price there are 4 missing data points




Based on the summary above, each column has 205 rows of data, seven columns containing missing data:
<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>

### Replace NaN values with mean value for column

In [11]:
# # getting average of a column
# avg_norm_loss = df["normalized-losses"].astype("float").mean(axis=0)
# print("Average of normalized-losses:", avg_norm_loss)
# # Replace "NaN" by mean value in "normalized-losses" column
# df["normalized-losses"].replace(np.nan, avg_norm_loss, inplace=True)

In [12]:
# above code made into function to apply to multiple columns
def replace_nan_with_mean(column):
    avg_of_column = (column).astype("float").mean(axis=0)
    print(f"Average of {column.name}:", avg_of_column)
    column.replace(np.nan, avg_of_column, inplace=True)
    print(f"Replaced all NaN values, in column {column.name}, with the average value of:", avg_of_column) 
    

In [13]:
# testing function
replace_nan_with_mean(df["normalized-losses"])

Average of normalized-losses: 122.0
Replaced all NaN values, in column normalized-losses, with the average value of: 122.0


In [14]:
# mass apply function to replace nan values with the average value of the column
column_list = df["normalized-losses"],df['bore'],df['stroke'],df['horsepower'],df['peak-rpm']
for x in column_list:
    replace_nan_with_mean(x)
    print('\n\n')

Average of normalized-losses: 122.0
Replaced all NaN values, in column normalized-losses, with the average value of: 122.0



Average of bore: 3.3297512437810957
Replaced all NaN values, in column bore, with the average value of: 3.3297512437810957



Average of stroke: 3.2554228855721337
Replaced all NaN values, in column stroke, with the average value of: 3.2554228855721337



Average of horsepower: 104.25615763546799
Replaced all NaN values, in column horsepower, with the average value of: 104.25615763546799



Average of peak-rpm: 5125.369458128079
Replaced all NaN values, in column peak-rpm, with the average value of: 5125.369458128079





### Replace NaN values with mode value for column

In [15]:
df['num-of-doors'].value_counts()


four    114
two      89
Name: num-of-doors, dtype: int64

In [16]:
# find the most frequent value (mode)
df['num-of-doors'].value_counts().idxmax()

#replace the missing 'num-of-doors' values by the most frequent 
df["num-of-doors"].replace(np.nan, "four", inplace=True)

# reset index, because we droped two rows
df.reset_index(drop=True, inplace=True)

df

symboling normalized-losses         make fuel-type aspiration  \
0            3               122  alfa-romero       gas        std   
1            3               122  alfa-romero       gas        std   
2            1               122  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

### Drop NaN value in target column

In [17]:
# drop all rows with NA values in price column (my target value)
df.dropna(subset=["price"], axis=0, inplace=True)

In [18]:
#double checking for missing values
find_missing_values(df)

## Convert data types to proper format

In [19]:
# Checking data types of the dataframe before change
data_types1 = df.dtypes
data_types1

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

In [20]:
# # Converting the incorrect data type to the correct types

# df[["bore", "stroke"]] = df[["bore", "stroke"]].astype("float")
# df[["normalized-losses"]] = df[["normalized-losses"]].astype("int")
# df[["price"]] = df[["price"]].astype("float")
# df[["peak-rpm"]] = df[["peak-rpm"]].astype("float")

### above code in function form  ###

def change_column_dtype(columns_to_transform,type):
    df[columns_to_transform] = df[columns_to_transform].astype(type)
    
change_column_dtype(["bore", "stroke","price","peak-rpm"],"float")
change_column_dtype(["normalized-losses"],"int64")
data_types2 = df.dtypes

In [21]:
# Comparing data types of the dataframe before and after change
Compare_dtypes = data_types1 != data_types2
Compare_dtypes

symboling            False
normalized-losses     True
make                 False
fuel-type            False
aspiration           False
num-of-doors         False
body-style           False
drive-wheels         False
engine-location      False
wheel-base           False
length               False
width                False
height               False
curb-weight          False
engine-type          False
num-of-cylinders     False
engine-size          False
fuel-system          False
bore                  True
stroke                True
compression-ratio    False
horsepower           False
peak-rpm              True
city-mpg             False
highway-mpg          False
price                 True
dtype: bool

In [22]:
compare_dtypes_df = pd.DataFrame(data = (data_types1,data_types2, Compare_dtypes))
compare_dtypes_df

symboling normalized-losses    make fuel-type aspiration num-of-doors  \
0     int64            object  object    object     object       object   
1     int64             int64  object    object     object       object   
2     False              True   False     False      False        False   

  body-style drive-wheels engine-location wheel-base  ... engine-size  \
0     object       object          object    float64  ...       int64   
1     object       object          object    float64  ...       int64   
2      False        False           False      False  ...       False   

  fuel-system     bore   stroke compression-ratio horsepower peak-rpm  \
0      object   object   object           float64     object   object   
1      object  float64  float64           float64     object  float64   
2       False     True     True             False      False     True   

  city-mpg highway-mpg    price  
0    int64       int64   object  
1    int64       int64  float64  
2    False       False     True  

[3 rows x 26 columns]

In [23]:
# transposing the dataframe (easier to read vertically)
compare_dtypes_table = compare_dtypes_df.T.reset_index()
compare_dtypes_table

index        0        1      2
0           symboling    int64    int64  False
1   normalized-losses   object    int64   True
2                make   object   object  False
3           fuel-type   object   object  False
4          aspiration   object   object  False
5        num-of-doors   object   object  False
6          body-style   object   object  False
7        drive-wheels   object   object  False
8     engine-location   object   object  False
9          wheel-base  float64  float64  False
10             length  float64  float64  False
11              width  float64  float64  False
12             height  float64  float64  False
13        curb-weight    int64    int64  False
14        engine-type   object   object  False
15   num-of-cylinders   object   object  False
16        engine-size    int64    int64  False
17        fuel-system   object   object  False
18               bore   object  float64   True
19             stroke   object  float64   True
20  compression-ratio  float64  float64  False
21         horsepower   object   object  False
22           peak-rpm   object  float64   True
23           city-mpg    int64    int64  False
24        highway-mpg    int64    int64  False
25              price   object  float64   True

In [24]:
columns=["Columns","Orignal dtype","Changed dtype", "Did dtype change?"]
compare_dtypes_table1  = compare_dtypes_table.rename(columns={'index':columns[0],0:columns[1],1:columns[2],2:columns[3]})
compare_dtypes_table1

Columns Orignal dtype Changed dtype Did dtype change?
0           symboling         int64         int64             False
1   normalized-losses        object         int64              True
2                make        object        object             False
3           fuel-type        object        object             False
4          aspiration        object        object             False
5        num-of-doors        object        object             False
6          body-style        object        object             False
7        drive-wheels        object        object             False
8     engine-location        object        object             False
9          wheel-base       float64       float64             False
10             length       float64       float64             False
11              width       float64       float64             False
12             height       float64       float64             False
13        curb-weight         int64         int64             False
14        engine-type        object        object             False
15   num-of-cylinders        object        object             False
16        engine-size         int64         int64             False
17        fuel-system        object        object             False
18               bore        object       float64              True
19             stroke        object       float64              True
20  compression-ratio       float64       float64             False
21         horsepower        object        object             False
22           peak-rpm        object       float64              True
23           city-mpg         int64         int64             False
24        highway-mpg         int64         int64             False
25              price        object       float64              True

## Data Normalization

In [25]:
# # replace (original value) with (original value)/(maximum value)
# df['length'] = df['length']/df['length'].max()
# df['width'] = df['width']/df['width'].max()
# df['height'] = df['height']/df['height'].max()

### above code in function form  ###

# scaling variable so the variable values range from 0 to 1
def normalize_column(columns_to_transform):
    for x in columns_to_transform:
        df[x] = df[x]/df[x].max()
    
    return df[columns_to_transform].head()
    
normalize_column(["length","width","height"])

length     width    height
0  0.811148  0.890278  0.816054
1  0.811148  0.890278  0.816054
2  0.822681  0.909722  0.876254
3  0.848630  0.919444  0.908027
4  0.848630  0.922222  0.908027

In [26]:
# saving data frame to csv
df.to_csv('clean_df.csv')

## Indicator variable (dummy variable)

transforming categorical values in a column to numerical variable

In [27]:
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

In [28]:
# get indicator variables and assign it to dummy variable
dummy_variable_1 = pd.get_dummies(df["fuel-type"])
dummy_variable_1.head()

diesel  gas
0       0    1
1       0    1
2       0    1
3       0    1
4       0    1

In [29]:
# change column names for ease of readability 
dummy_variable_1.rename(columns={'fuel-type-diesel':'gas', 'fuel-type-diesel':'diesel'}, inplace=True)
dummy_variable_1.head()

diesel  gas
0       0    1
1       0    1
2       0    1
3       0    1
4       0    1

In [30]:
# merge data frame "df" and "dummy_variable_1" 
df = pd.concat([df, dummy_variable_1], axis=1)

# drop original column "fuel-type" from "df"
df.drop("fuel-type", axis = 1, inplace=True)
df.head()

symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   
3          2                164         audi        std         four   
4          2                164         audi        std         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  bore  \
0  convertible          rwd           front        88.6  0.811148  ...  3.47   
1  convertible          rwd           front        88.6  0.811148  ...  3.47   
2    hatchback          rwd           front        94.5  0.822681  ...  2.68   
3        sedan          fwd           front        99.8  0.848630  ...  3.19   
4        sedan          4wd           front        99.4  0.848630  ...  3.19   

   stroke  compression-ratio horsepower peak-rpm  city-mpg highway-mpg  \
0    2.68                9.0        111   5000.0        21          27   
1    2.68                9.0        111   5000.0        21          27   
2    3.47                9.0        154   5000.0        19          26   
3    3.40               10.0        102   5500.0        24          30   
4    3.40                8.0        115   5500.0        18          22   

     price  diesel  gas  
0  13495.0       0    1  
1  16500.0       0    1  
2  16500.0       0    1  
3  13950.0       0    1  
4  17450.0       0    1  

[5 rows x 27 columns]

In [31]:
# get indicator variables of aspiration and assign it to data frame "dummy_variable_2"
dummy_variable_2 = pd.get_dummies(df['aspiration'])

# change column names for clarity
dummy_variable_2.rename(columns={'std':'aspiration-std', 'turbo': 'aspiration-turbo'}, inplace=True)

#merge the new dataframe to the original datafram
df = pd.concat([df, dummy_variable_2], axis=1)

# drop original column "aspiration" from "df"
df.drop('aspiration', axis = 1, inplace=True)

In [32]:
# get indicator variables of num-of-doors and assign it to data frame "dummy_variable_3"
dummy_variable_3 = pd.get_dummies(df['num-of-doors'])

# change column names for clarity
dummy_variable_3.rename(columns={'two':'num-of-doors-two', 'four': 'num-of-doors-four'}, inplace=True)

#merge the new dataframe to the original datafram
df = pd.concat([df, dummy_variable_3], axis=1)

# drop original column "num-of-doors" from "df"
df.drop('num-of-doors', axis = 1, inplace=True)


In [33]:
# get indicator variables of engine-location and assign it to data frame "dummy_variable_4"
dummy_variable_4 = pd.get_dummies(df['engine-location'])

# change column names for clarity
dummy_variable_4.rename(columns={'front':'engine-location-front', 'back': 'engine-location-back'}, inplace=True)

#merge the new dataframe to the original datafram
df = pd.concat([df, dummy_variable_4], axis=1)

# drop original column "engine-location" from "df"
df.drop('engine-location', axis = 1, inplace=True)


In [34]:
# get indicator variables of drive-wheels and assign it to data frame "dummy_variable_5"
dummy_variable_5 = pd.get_dummies(df['drive-wheels'])

# change column names for clarity
dummy_variable_5.rename(columns={'rwd':'drive-wheels-rwd', 'fwd': 'drive-wheels-fwd', '4wd': 'drive-wheels-4wd'}, inplace=True)

#merge the new dataframe to the original datafram
df = pd.concat([df, dummy_variable_5], axis=1)

# drop original column "drive-wheels" from "df"
df.drop('drive-wheels', axis = 1, inplace=True)


In [35]:
# get indicator variables of num-of-cylinders and assign it to data frame "dummy_variable_6"
dummy_variable_6 = pd.get_dummies(df['num-of-cylinders'])

# change column names for clarity
dummy_variable_6.rename(columns={'two':'num-of-cylinders-two', 'three': 'num-of-cylinders-three', 'four': 'num-of-cylinders-four', 'five': 'num-of-cylinders-five', 'six': 'num-of-cylinders-six', 'eight': 'num-of-cylinders-eight', 'twelve': 'num-of-cylinders-twelve'}, inplace=True)

#merge the new dataframe to the original datafram
df = pd.concat([df, dummy_variable_6], axis=1)

# drop original column "num-of-cylinders" from "df"
df.drop('num-of-cylinders', axis = 1, inplace=True)


In [36]:
df

symboling  normalized-losses         make   body-style  wheel-base  \
0            3                122  alfa-romero  convertible        88.6   
1            3                122  alfa-romero  convertible        88.6   
2            1                122  alfa-romero    hatchback        94.5   
3            2                164         audi        sedan        99.8   
4            2                164         audi        sedan        99.4   
..         ...                ...          ...          ...         ...   
200         -1                 95        volvo        sedan       109.1   
201         -1                 95        volvo        sedan       109.1   
202         -1                 95        volvo        sedan       109.1   
203         -1                 95        volvo        sedan       109.1   
204         -1                 95        volvo        sedan       109.1   

       length     width    height  curb-weight engine-type  ...  \
0    0.811148  0.890278  0.816054         2548        dohc  ...   
1    0.811148  0.890278  0.816054         2548        dohc  ...   
2    0.822681  0.909722  0.876254         2823        ohcv  ...   
3    0.848630  0.919444  0.908027         2337         ohc  ...   
4    0.848630  0.922222  0.908027         2824         ohc  ...   
..        ...       ...       ...          ...         ...  ...   
200  0.907256  0.956944  0.928094         2952         ohc  ...   
201  0.907256  0.955556  0.928094         3049         ohc  ...   
202  0.907256  0.956944  0.928094         3012        ohcv  ...   
203  0.907256  0.956944  0.928094         3217         ohc  ...   
204  0.907256  0.956944  0.928094         3062         ohc  ...   

     drive-wheels-4wd drive-wheels-fwd  drive-wheels-rwd  \
0                   0                0                 1   
1                   0                0                 1   
2                   0                0                 1   
3                   0                1                 0   
4                   1                0                 0   
..                ...              ...               ...   
200                 0                0                 1   
201                 0                0                 1   
202                 0                0                 1   
203                 0                0                 1   
204                 0                0                 1   

     num-of-cylinders-eight  num-of-cylinders-five num-of-cylinders-four  \
0                         0                      0                     1   
1                         0                      0                     1   
2                         0                      0                     0   
3                         0                      0                     1   
4                         0                      1                     0   
..                      ...                    ...                   ...   
200                       0                      0                     1   
201                       0                      0                     1   
202                       0                      0                     0   
203                       0                      0                     0   
204                       0                      0                     1   

     num-of-cylinders-six  num-of-cylinders-three  num-of-cylinders-twelve  \
0                       0                       0                        0   
1                       0                       0                        0   
2                       1                       0                        0   
3                       0                       0                        0   
4                       0                       0                        0   
..                    ...                     ...                      ...   
200                     0                       0                        0   
201                     0                

In [37]:
# saving data frame to csv
df.to_csv('clean_df_with_dummy.csv')